### ISSUE TITLE
Loss value discrepancy between SP and non-SP modes with balance_dp_token=True

### ISSUE THREAD

**Issue by wenxueru** [verl-project/verl#1963]
There might be a discrepancy in initial loss values when comparing training with sequence parallelism (SP) enabled versus disabled, specifically when `balance_dp_token=True` is set.

In the following code:
```python
if self.config.data.balance_dp_token:
    torch.distributed.all_reduce(valid_token_this_rank)
    dp_size = self.ulysses_device_mesh.size(0) if use_sp else torch.distributed.get_world_size()
else:
    dp_size = 1

loss = torch.sum(loss) / (valid_token_this_rank + 1e-8) * dp_size
```

When `balance_dp_token=True`, initial loss values differ by ~SP size factor between SP and non-SP modes:
- With SP=2: Uses data parallel dimension (2) as multiplier
- Without SP: Uses total process count (4) as multiplier
- Result: ~2x difference

Maybe just use world_size for avoiding confusion?

**Comment by vermouth1992** [comment]
You are right. I guess all reduce should only be performed inside dp group, not the whole world.

**Comment by wenxueru** [comment]
> You are right. I guess all reduce should only be performed inside dp group, not the whole world.

Will the ranks within the same SP have the same valid_token_this_rank?

**Comment by zui-jiang** [comment]
> > You are right. I guess all reduce should only be performed inside dp group, not the whole world.
> 
> Will the ranks within the same SP have the same valid_token_this_rank?


I think in the same SP group,  each node contains a chunk of the prompt and response along the sequence dimension.

**Comment by vermouth1992** [comment]
> > You are right. I guess all reduce should only be performed inside dp group, not the whole world.
> 
> Will the ranks within the same SP have the same valid_token_this_rank?

Yes. But the last rank may contain paddings.

**Comment by Kirrito-k423** [comment]
### 中文
根据本次定时任务对 comments 和 timeline 的检查，这个 issue 命中默认超时关闭规则：最近一次非当前运行账号（Kirrito-k423）的公开回复或公开引用是 2025-06-12T14:28:20Z，已经超过 6 个月。证据：latest_non_self_reply_or_reference_at=2025-06-12T14:28:20Z; kind=comment; actor=vermouth1992; reference=https://github.com/verl-project/verl/issues/1963#issuecomment-2967055859。检查来源：https://api.github.com/repos/verl-project/verl/issues/1963/comments 和 https://api.github.com/repos/verl-project/verl/issues/1963/timeline。因此按定时清理标准先关闭；如果作者或社区成员仍能提供新的复现、证据或需求背景，可以继续评论请求重开或创建新 issue。

### English
In this scheduled pass, I checked the issue comments and timeline. This issue matches the default stale-issue closure rule: the latest public reply or public reference from an account other than the current runner (Kirrito-k423) was at 2025-06-12T14:28:20Z, which is more than 6 months ago. Evidence: latest_non_self_reply_or_reference_at=2025-06-12T14:28:20Z; kind=comment; actor=vermouth1992; reference=https://github.com/verl-project/verl/issues/1963#issuecomment-2967055859. Checked sources: https://api.github.com/repos/verl-project/verl/issues/1963/comments and https://api.github.com/repos/verl-project/verl/issues/1963/timeline. I am closing it under the scheduled cleanup rule. If the author or community can provide a fresh reproduction, evidence, or requirement context, please comment to request reopening or open a new issue.

This response was generated by a scheduled Codex GPT5.5 issue processing task.
